docs: tcp support (#1712)
pomerium/pomerium: docs/docs/topics/tcp-support.md

---
title: TCP Support
description: >-

>   This article describes how to leverage pomerium for TCP proxying

Reviewer: keywords (ssh, tcp, etc)

Author: 👍

---

# TCP Support

## About

> Internal applications are often HTTP, but the engineering teams that support those applications frequently require access to lower level administrative and data protcols such as SSH, RDP, Postgres, MySQL, Redis, etc.

Reviewer: ```suggestion
Operations and engineering teams frequently require access to lower level administrative and data protocols such as SSH, RDP, Postgres, MySQL, Redis, etc.
```

take or leave it

Reviewer: n.b. either way "protocols"

> Though Pomerium is primarily designed to authorize and proxy HTTP traffic, it can also be leveraged to protect these non-HTTP systems with the same consistent authorization policy.  This is achieved by tunneling TCP over HTTP with the help of a client side command built into [`pomerium-cli`](/docs/installation.md#pomerium-cli).

Reviewer: ```suggestion
In addition to managing HTTP based applications, Pomerium can be used to protect non-HTTP systems with the same consistent authorization policy. This is achieved by tunneling TCP over HTTP with the help of a client side command built into [`pomerium-cli`](/docs/installation.md#pomerium-cli).
```

take it or leave it

Internally, Pomerium uses the [`CONNECT` method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT) to establish the TCP tunnel.

::: warning

> To minimize issues with TCP support, Pomerium should not be placed behind another HTTP proxy.  Instead, configure your load balancer in L4 or TCP mode.

Author: @desimone I took a fresh pass at this.

Reviewer: LGTM!

> For proper TCP support, you should not use an HTTP proxy in front of Pomerium.  Instead, configure your load balancer in L4 or TCP mode.

Reviewer: ```suggestion
To ensure TCP support, Pomerium should not be used behind another HTTP proxy.  Instead, configure your load balancer in L4 or TCP passthrough mode.
```

> If an HTTP proxy is in front of Pomerium, it must know how to properly handle the `CONNECT` command and proxy it upstream.  This will be specific to each proxy implementation.

Reviewer: ```suggestion
Otherwise, the HTTP proxy fronting Pomerium must know how to properly handle the `CONNECT` command and proxy it upstream.  This will be specific to each proxy implementation.
```

:::

## Configuring

> TCP configuration is very simple to configure - just specify the correct scheme and ports in your policy [`to`](/reference/#to) and [`from`](/reference/#from) fields.

Reviewer: ```suggestion
TCP configuration is simple. Just specify the correct scheme and ports in your policy [`to`](/reference/#to) and [`from`](/reference/#from) fields.
```

Example:
```yaml
policy:
    - from: tcp+https://redis.corp.example.com:6379
      to: tcp://redis.internal.example.com:6379
      allowed_users:
        - contractor@not-example.com
      allowed_groups:
        - datascience@example.com
```

Notes:

* When configuring a TCP route, any HTTP specific settings such as `regex_rewrite_pattern`, or `set_request_headers` have no effect.
* While data is encrypted from a user system to Pomerium's proxy, the underlying application protocol must also support encryption for data to be fully encrypted end-to-end.
* The ports in `from` and `to` are independent.  Users only need to know the `from` URL to connect.  The `to` can be changed without end user participation.

## Using

> While HTTP routes can be consumed with just a normal browswer, `pomerium-cli` must serve as a proxy for TCP routes.  It is [available](/docs/installation.md#pomerium-cli) for a variety of platforms in various formats.

Reviewer: ```suggestion
While HTTP routes can be consumed with just a normal browser, `pomerium-cli` must serve as a proxy for TCP routes.  It is [available](/docs/installation.md#pomerium-cli) for a variety of platforms in various formats.
```

To connect, you normally need just the external hostname and port of your TCP route:

```bash
% pomerium-cli tcp redis.corp.example.com:6379
5:57PM INF tcptunnel: listening on 127.0.0.1:52046
```

By default, `pomerium-cli` will start a listener on loopback on a random port.

> On first connection, you will be sent through a standard Pomerium HTTP authentiation flow.  After completing this, your TCP connection should be established!

Reviewer: ```suggestion
On first connection, you will be sent through a standard Pomerium HTTP authentication flow.  After completing this, your TCP connection should be established!
```
